File: component/mix/mix.go
// Package mix implements the `mix` and `dynamic-mix` components: the
// simplest mixers, summing every input wire and multiplying by a single
// level scale. Mirrors ondes.synth.wire.Junction and DynamicJunction.
//
// `dynamic-mix` additionally tracks MIDI controller 7 (channel volume),
// setting level-scale = value/128. The regression MIDI carries no CC
// data, so for those fixtures a dynamic-mix behaves identically to a
// plain mix (level-scale 1).
package mix

import (
	"ondes/component"
	"ondes/synth"
)

func init() {
	component.Register("mix", func() component.Component { return &Mix{} })
	component.Register("dynamic-mix", func() component.Component { return &Mix{dynamic: true} })
}

// Mix sums its inputs and scales the result. When dynamic, it responds
// to MIDI volume (CC 7) like Java's DynamicJunction.
type Mix struct {
	out        *synth.Wire
	inputs     []*synth.Wire
	levelScale float64
	dynamic    bool
}

func (m *Mix) Configure(spec component.Spec, v *synth.Voice, _ string) error {
	m.levelScale = 1
	if ls, ok := numeric(spec["level-scale"]); ok {
		m.levelScale = ls
	}
	m.out = v.NewWire(m.compute)
	return nil
}

func (m *Mix) Output() *synth.Wire { return m.out }

func (m *Mix) AddInput(_ string, w *synth.Wire) { m.inputs = append(m.inputs, w) }

func (m *Mix) compute() float64 {
	var s float64
	for _, w := range m.inputs {
		s += w.Get()
	}
	return s * m.levelScale
}

// OnMidi implements the MIDI volume response of a dynamic-mix. Plain
// mixes ignore it.
func (m *Mix) OnMidi(msg synth.MidiMsg) {
	if !m.dynamic {
		return
	}
	// Controller change, CC 7 (channel volume).
	if msg.Status&0xF0 == 0xB0 && msg.Data1 == 7 {
		m.levelScale = float64(msg.Data2) / 128.0
	}
}

func numeric(v any) (float64, bool) {
	switch x := v.(type) {
	case int:
		return float64(x), true
	case int64:
		return float64(x), true
	case float64:
		return x, true
	}
	return 0, false
}

File: c/src/component/mix.c
// `mix` and `dynamic-mix`: sum every input wire and multiply by a single
// level scale. `dynamic-mix` additionally tracks MIDI CC 7 (channel
// volume). Ported from component/mix/mix.go.
#include "computil.h"
#include "ondes/component.h"

typedef struct {
    Component base;
    Wire *out;
    WireList inputs;
    double level_scale;
    bool dynamic;
} Mix;

static double mix_compute(void *ctx) {
    Mix *m = ctx;
    return wirelist_sum(&m->inputs) * m->level_scale;
}

static Wire *mix_output(Component *self) { return ((Mix *)self)->out; }

static void mix_add_input(Component *self, const char *sel, Wire *w) {
    (void)sel;
    wirelist_add(&((Mix *)self)->inputs, w);
}

static void mix_on_midi(Component *self, MidiMsg m) {
    Mix *mix = (Mix *)self;
    if (!mix->dynamic) return;
    if ((m.status & 0xF0) == 0xB0 && m.data1 == 7)
        mix->level_scale = (double)m.data2 / 128.0;
}

static int mix_configure(Component *self, const Spec *spec, Voice *v,
                         const char *name) {
    (void)name;
    Mix *m = (Mix *)self;
    wirelist_init(&m->inputs, voice_arena(v));
    m->level_scale = spec_double(spec, "level-scale", 1.0);
    m->out = voice_new_wire(v, mix_compute, m);
    return 0;
}

static const ComponentVTable MIX_VT = {
    .configure = mix_configure,
    .output = mix_output,
    .add_input = mix_add_input,
    .on_midi = mix_on_midi,
};

Component *mix_new(Arena *a) {
    Mix *m = arena_alloc(a, sizeof(*m));
    m->base.vt = &MIX_VT;
    return &m->base;
}

Component *dynamic_mix_new(Arena *a) {
    Mix *m = arena_alloc(a, sizeof(*m));
    m->base.vt = &MIX_VT;
    m->dynamic = true;
    return &m->base;
}
